### PR TITLE
fix(core): fix non-partitioned table corruption caused by `truncate`  & table corruption after restore from snapshot

### PR DESCRIPTION
**Pull request by bluestreak01**
It all started when I saw flapping test on one of the PRs..... it escalated quickly

### Fixed two issues

- non-partitioned table will be corrupt after `update` followed by `truncate`, we reset metada in the way we should not have
- after snapshot some tables may be corrupt if column was removed during the snapshot. This was caused by "no-op" database snapshot agent being injected quite randomly into the writer pool.

### Lessons learned

- never ignore the result of `Files.remove()` call
- try to avoid leavng test code (such as no-op snapshot agent) in the production branch, move things to test branch as much as possible

### Additional refactoring

- refactored out no-op agent so it cannot be used again
- `FilesFacade.remove()` method is "loud", it will throw an exception if file cannot be removed, there is `FilesFacade.removeQuiet()` for cases when exception is undesireable

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 92 / 108 (85.19%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/CairoEngine.java|16|26|61.54%|
|:large_blue_circle:|io/questdb/cairo/IndexBuilder.java|3|4|75.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|29|34|85.29%|
|:large_blue_circle:|io/questdb/cairo/TableConverter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalWriter.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/log/LogRollingFileWriter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/PurgingOperator.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/cairo/DatabaseSnapshotAgentImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/FilesFacadeImpl.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/ApplyWal2TableJob.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompilerImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/ColumnVersionWriter.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/cairo/RebuildColumnBase.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/DropIndexOperator.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/ColumnPurgeOperator.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/text/CopyTask.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableNameRegistryStore.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 92 / 108 (85.19%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/CairoEngine.java|16|26|61.54%|
|:large_blue_circle:|io/questdb/cairo/IndexBuilder.java|3|4|75.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|29|34|85.29%|
|:large_blue_circle:|io/questdb/cairo/TableConverter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalWriter.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/log/LogRollingFileWriter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/PurgingOperator.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/cairo/DatabaseSnapshotAgentImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/FilesFacadeImpl.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/ApplyWal2TableJob.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompilerImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/ColumnVersionWriter.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/cairo/RebuildColumnBase.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/DropIndexOperator.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/ColumnPurgeOperator.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/text/CopyTask.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableNameRegistryStore.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 93 / 108 (86.11%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/CairoEngine.java|17|26|65.38%|
|:large_blue_circle:|io/questdb/cairo/IndexBuilder.java|3|4|75.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|29|34|85.29%|
|:large_blue_circle:|io/questdb/cairo/TableConverter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalWriter.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/log/LogRollingFileWriter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/PurgingOperator.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/cairo/DatabaseSnapshotAgentImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/FilesFacadeImpl.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/ApplyWal2TableJob.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompilerImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/ColumnVersionWriter.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/cairo/RebuildColumnBase.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/DropIndexOperator.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/ColumnPurgeOperator.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/text/CopyTask.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableNameRegistryStore.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 99 / 115 (86.09%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/CairoEngine.java|17|26|65.38%|
|:large_blue_circle:|io/questdb/cairo/IndexBuilder.java|3|4|75.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|29|34|85.29%|
|:large_blue_circle:|io/questdb/cairo/wal/ApplyWal2TableJob.java|9|10|90.00%|
|:large_blue_circle:|io/questdb/cairo/TableConverter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalWriter.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/log/LogRollingFileWriter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/PurgingOperator.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/cairo/DatabaseSnapshotAgentImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/FilesFacadeImpl.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompilerImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/ColumnVersionWriter.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/cairo/RebuildColumnBase.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/DropIndexOperator.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/ColumnPurgeOperator.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/text/CopyTask.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableNameRegistryStore.java|1|1|100.00%|

File: core/src/main/resources/io/questdb/bin/windows/libquestdb.dll
Comment (ideoma) [inline core/src/main/resources/io/questdb/bin/windows/libquestdb.dll:1]: Are there any changes in the native code? I don't see any, probably should be reverted.

File: core/src/main/java/io/questdb/cairo/wal/ApplyWal2TableJob.java [hunk 1]
```diff
@@ -146,6 +144,14 @@ private static boolean cleanDroppedTableDirectory(CairoEngine engine, Path tempP
         return false;
     }
 
+    private static boolean removeOrLog(Path path, FilesFacade ff) {
+        if (!ff.removeQuiet(path.$())) {
+            LOG.info().$("could not remove, will retry [path=").$(path).$(", errno=").$(ff.errno()).I$();
```
Comment (ideoma) [inline core/src/main/java/io/questdb/cairo/wal/ApplyWal2TableJob.java:149]: My bad, should be
```suggestion
            LOG.info().$("could not remove, will retry [path=").utf8(path).$(", errno=").$(ff.errno()).I$();
```

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 99 / 115 (86.09%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/CairoEngine.java|17|26|65.38%|
|:large_blue_circle:|io/questdb/cairo/IndexBuilder.java|3|4|75.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|29|34|85.29%|
|:large_blue_circle:|io/questdb/cairo/wal/ApplyWal2TableJob.java|9|10|90.00%|
|:large_blue_circle:|io/questdb/cairo/TableConverter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalWriter.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/log/LogRollingFileWriter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/PurgingOperator.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/cairo/DatabaseSnapshotAgentImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/FilesFacadeImpl.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompilerImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/ColumnVersionWriter.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/cairo/RebuildColumnBase.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/DropIndexOperator.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/ColumnPurgeOperator.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/text/CopyTask.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableNameRegistryStore.java|1|1|100.00%|

File: core/src/main/java/io/questdb/cairo/CairoEngine.java [hunk 2]
```diff
@@ -1073,10 +1084,49 @@ public void unlockReaders(TableToken tableToken) {
...
+
     public void unlockTableName(TableToken tableToken) {
         tableNameRegistry.unlockTableName(tableToken);
     }
 
+    public long update(CharSequence updateSql, SqlExecutionContext sqlExecutionContext) throws SqlException {
+        return update(updateSql, sqlExecutionContext, null);
+    }
+
+    public long update(CharSequence updateSql, SqlExecutionContext sqlExecutionContext, @Nullable SCSequence eventSubSeq) throws SqlException {
```
[the remark this answers is not in the document]
Comment (bluestreak01) [reply core/src/main/java/io/questdb/cairo/CairoEngine.java:1118]: yep, it is intentional. All other similar methods (`insert`, `ddl`, `select`) are on the engine, `update` was missing. The idea to make this methof available to whoever needs to execute update, such as embedded user or server (i did not refactor server)

**Review by ideoma — APPROVED** [review]

**Pull request merged**